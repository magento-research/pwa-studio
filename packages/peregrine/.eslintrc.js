const config = {
    parser: 'babel-eslint',
    extends: ['@magento'],
    rules: {
        "no-undef": "off"
    }
};

module.exports = config;
