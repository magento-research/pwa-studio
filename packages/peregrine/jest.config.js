module.exports = {
    displayName: 'Peregrine',
    setupFiles: ['<rootDir>/scripts/shim.js', '<rootDir>/scripts/fetch-mock.js']
};
