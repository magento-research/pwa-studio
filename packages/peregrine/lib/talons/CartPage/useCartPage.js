import { useCallback, useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';
import { useApolloClient, useLazyQuery, useMutation } from '@apollo/react-hooks';

import { useAppContext } from '@magento/peregrine/lib/context/app';
import { useUserContext } from '@magento/peregrine/lib/context/user';
import { useCartContext } from '@magento/peregrine/lib/context/cart';

export const useCartPage = props => {
    const {
        queries: { getCartDetails },
        mutations: { signOutMutation }
    } = props;

    const { resetStore } = useApolloClient();
    const history = useHistory();

    const [, { toggleDrawer }] = useAppContext();
    const [{ isSignedIn }, { signOut }] = useUserContext();
    const [{ cartId }] = useCartContext();

    const [isCartUpdating, setIsCartUpdating] = useState(false);

    const [revokeToken] = useMutation(signOutMutation);
    const [fetchCartData, { data }] = useLazyQuery(getCartDetails, {
        // TODO: Purposely overfetch and hit the network until all components
        // are correctly updating the cache. Will be fixed by PWA-321.
        fetchPolicy: 'cache-and-network'
    });

    const handleSignInToggle = useCallback(async () => {
        if (isSignedIn) {
            // After logout, reset the store to set the bearer token.
            // https://www.apollographql.com/docs/react/networking/authentication/#reset-store-on-logout
            await resetStore();
            await signOut({ revokeToken });

            // Refresh this page.
            history.go(0);
        }
        else {
            toggleDrawer('nav');
        }
    }, [history, isSignedIn, resetStore, revokeToken, signOut, toggleDrawer]);

    useEffect(() => {
        if (cartId) {
            fetchCartData({
                variables: {
                    cartId
                }
            });
        }
    }, [cartId, fetchCartData]);

    return {
        hasItems: !!(data && data.cart.total_quantity),
        handleSignInToggle,
        isSignedIn,
        isCartUpdating,
        setIsCartUpdating
    };
};
