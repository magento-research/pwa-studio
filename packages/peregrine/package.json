{
  "name": "@magento/peregrine",
  "version": "3.0.0",
  "publishConfig": {
    "access": "public"
  },
  "description": "The core runtime of Magento PWA",
  "main": "lib/index.js",
  "scripts": {
    "build": " ",
    "build:ci": " ",
    "build:cjs": " ",
    "build:dev": " ",
    "clean": " ",
    "storybook": "start-storybook -p 9001 -c .storybook",
    "storybook:build": "build-storybook -c .storybook -o storybook-dist",
    "watch": " "
  },
  "repository": "github:magento-research/pwa-studio",
  "author": "Magento Commerce",
  "license": "(OSL-3.0 OR AFL-3.0)",
  "bugs": {
    "url": "https://github.com/magento-research/pwa-studio/issues"
  },
  "homepage": "https://github.com/magento-research/pwa-studio/tree/master/packages/peregrine#readme",
  "devDependencies": {
    "@storybook/react": "~4.1.13",
    "intl": "~1.2.5",
    "react": "~16.8.3",
    "react-apollo": "~2.5.3",
    "react-dom": "~16.8.3",
    "react-router-dom": "~5.0.0",
    "react-test-renderer": "~16.8.6",
    "rimraf": "~2.6.3"
  },
  "peerDependencies": {
    "@babel/runtime": "^7.4.2",
    "react": "^16.8.0",
    "react-apollo": "^2.5.3",
    "react-dom": "^16.8.0",
    "react-router-dom": "~5.0.0"
  },
  "engines": {
    "node": "8.10.0 || >=10.7.0",
    "yarn": ">=1.12.0"
  },
  "module": "lib/index.js",
  "jsnext:main": "lib/index.js",
  "es2015": "lib/index.js",
  "sideEffects": false
}
