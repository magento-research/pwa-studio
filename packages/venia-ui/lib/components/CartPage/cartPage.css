.root {
    padding: 1.5rem;
}

.body {
    display: grid;
    gap: 2rem;
    grid-template-columns: 66% auto;
    grid-template-areas:
        'items summary'
        'adjustments summary'
        'recents summary';
}

.heading_container {
    display: flex;
    align-items: baseline;
    justify-content: space-between;

    margin-bottom: 1rem;
    margin-top: 1rem;
}

.heading {
}

.items_container {
    grid-area: items;
}

.price_adjustments_container {
    grid-area: adjustments;
}

.recently_viewed_container {
    grid-area: recents;
}

.sign_in {
    font-weight: bold;
    text-decoration: underline;
}

.summary_container {
    grid-area: summary;
    height: 100%;
}

.summary_contents {
    position: sticky;
    /*
     * TODO: Use CSS Properties (variables) or something instead of hardcoding these. 
     * min-height of nav header + page spacing + height of page header 
     */
    top: calc(3.5rem + 3.5rem + 25px);
}

/*
 * Mobile-specific styles.
 */

@media (max-width: 640px) {
    .body {
        /* Only one column in mobile view. */
        grid-template-columns: 100%;
        grid-template-areas:
            'items'
            'adjustments'
            'summary';
    }

    .recently_viewed_container {
        display: none;
    }
}
