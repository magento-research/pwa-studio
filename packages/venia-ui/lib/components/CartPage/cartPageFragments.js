import gql from 'graphql-tag';

export const CartPageFragment = gql`
    fragment CartPageFragment on Cart {
        total_quantity
    }
`;
