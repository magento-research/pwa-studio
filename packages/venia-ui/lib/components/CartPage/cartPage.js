import React from 'react';

import { useCartPage } from '@magento/peregrine/lib/talons/CartPage/useCartPage';

import Button from '../Button';

import { mergeClasses } from '../../classify';
import defaultClasses from './cartPage.css';

const CartPage = props => {

    const { handleSignIn, isSignedIn } = useCartPage();

    const classes = mergeClasses(defaultClasses, props.classes);

    const signInDisplay = !isSignedIn ? (
        <Button
            className={classes.sign_in}
            onClick={handleSignIn}
            priority="high"
        >
            {'Sign In'}
        </Button>
    ) : null;

    return (
        <div className={classes.root}>
            <div className={classes.heading_container}>
                <h1 className={classes.heading}>Cart</h1>
                { signInDisplay }
            </div>
            <div className={classes.items_container}>
                <a href="https://jira.corp.magento.com/browse/PWA-238">Items List to be completed by PWA-238</a>.
            </div>
            <div className={classes.price_adjustments_container}>
                <a href="https://jira.corp.magento.com/browse/PWA-241">Price Adjustments to be completed by PWA-241</a>.
            </div>
            <div className={classes.summary_container}>
                <a href="https://jira.corp.magento.com/browse/PWA-240">Order Summary to be completed by PWA-240</a>.
            </div>
            <div className={classes.recently_viewed_container}>
                Recently Viewed TBD. Should show only in Desktop.
            </div>
        </div>
    );
}

export default CartPage;
