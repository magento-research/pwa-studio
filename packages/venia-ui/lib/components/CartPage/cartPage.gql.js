import gql from 'graphql-tag';
import { CartPageFragment } from './cartPageFragments.gql';

export const GET_CART_DETAILS = gql`
    query getCartDetails($cartId: String!) {
        cart(cart_id: $cartId) {
            id
            ...CartPageFragment
        }
    }
    ${CartPageFragment}
`;

export const SIGN_OUT = gql`
    mutation signOut {
        revokeCustomerToken {
            result
        }
    }
`;
