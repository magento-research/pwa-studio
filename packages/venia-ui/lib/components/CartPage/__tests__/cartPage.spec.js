import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import CartPage from '../cartPage';
import { useLazyQuery } from '@apollo/react-hooks';

jest.mock('@apollo/react-hooks', () => {
    const runQuery = jest.fn();
    const queryResult = {
        data: null,
        error: null,
        loading: false
    };
    const useLazyQuery = jest.fn(() => [runQuery, queryResult]);

    return { useLazyQuery };
});

jest.mock('@magento/peregrine/lib/context/app', () => {
    const api = {
        toggleDrawer: jest.fn()
    };
    const state = {};
    const useAppContext = jest.fn(() => [state, api]);

    return { useAppContext };
});

jest.mock('@magento/peregrine/lib/context/cart', () => {
    const api = {};
    const state = { cartId: 'cart123' };
    const useCartContext = jest.fn(() => [state, api]);

    return { useCartContext };
});

jest.mock('@magento/peregrine/lib/context/user', () => {
    const api = {};
    const state = { isSignedIn: false };
    const useUserContext = jest.fn(() => [state, api]);

    return { useUserContext };
});

test('renders no price adjustments or summary components if cart is empty', () => {
    useLazyQuery.mockReturnValueOnce([
        jest.fn(),
        {
            data: {
                cart: {
                    total_quantity: 0
                }
            }
        }
    ]);

    const instance = createTestInstance(<CartPage />);
    expect(instance.toJSON()).toMatchSnapshot();
});

test('renders price adjustments and summary components if cart has items', () => {
    useLazyQuery.mockReturnValueOnce([
        jest.fn(),
        {
            data: {
                cart: {
                    total_quantity: 1
                }
            }
        }
    ]);

    const instance = createTestInstance(<CartPage />);
    expect(instance.toJSON()).toMatchSnapshot();
});
