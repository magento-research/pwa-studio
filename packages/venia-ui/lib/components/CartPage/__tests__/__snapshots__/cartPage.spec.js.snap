// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders no price adjustments or summary components if cart is empty 1`] = `
<div>
  <div>
    <h1>
      Cart
    </h1>
    <button
      onClick={[Function]}
      type="button"
    >
      <span>
        Sign In
      </span>
    </button>
  </div>
  <div>
    <div>
      <h3>
        There are no items in your cart.
      </h3>
    </div>
    <div />
    <div>
      <div />
    </div>
    <div>
      <a
        href="https://jira.corp.magento.com/browse/PWA-270"
      >
        Recently Viewed to be completed by PWA-270.
      </a>
    </div>
  </div>
</div>
`;

exports[`renders price adjustments and summary components if cart has items 1`] = `
<div>
  <div>
    <h1>
      Cart
    </h1>
    <button
      onClick={[Function]}
      type="button"
    >
      <span>
        Sign In
      </span>
    </button>
  </div>
  <div>
    <div>
      <h3>
        There are no items in your cart.
      </h3>
    </div>
    <div>
      <div>
        <div>
          <div>
            <button
              onClick={[Function]}
            >
              <div>
                Select Shipping Method
              </div>
              <span>
                <svg
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polyline
                    points="18 15 12 9 6 15"
                  />
                </svg>
              </span>
            </button>
            <div>
              <a
                href="https://jira.corp.magento.com/browse/PWA-239"
              >
                Shipping Methods to be completed by PWA-239.
              </a>
            </div>
          </div>
          <div>
            <button
              onClick={[Function]}
            >
              <div>
                Enter Coupon Code
              </div>
              <span>
                <svg
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polyline
                    points="6 9 12 15 18 9"
                  />
                </svg>
              </span>
            </button>
            <div />
          </div>
          <div>
            <button
              onClick={[Function]}
            >
              <div>
                Apply Gift Card
              </div>
              <span>
                <svg
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polyline
                    points="6 9 12 15 18 9"
                  />
                </svg>
              </span>
            </button>
            <div />
          </div>
          <div>
            <button
              onClick={[Function]}
            >
              <div>
                See Gift Options
              </div>
              <span>
                <svg
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polyline
                    points="6 9 12 15 18 9"
                  />
                </svg>
              </span>
            </button>
            <div />
          </div>
        </div>
      </div>
    </div>
    <div>
      <div />
    </div>
    <div>
      <a
        href="https://jira.corp.magento.com/browse/PWA-270"
      >
        Recently Viewed to be completed by PWA-270.
      </a>
    </div>
  </div>
</div>
`;
